contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,318 @@
+
+        withdrawId = nextWithdrawId++;
+
+        withdrawals[withdrawId] = WithdrawRequest(
+            validatorId,
+            undelegateAmount
+        );
+        pendingWithdrawals += undelegateAmount;
+
+        ISFC(sfc).undelegate(validatorId, withdrawId, undelegateAmount);
+
+        emit Undelegated(withdrawId, validatorId, undelegateAmount);
+    }
+
+    // slither-disable-start reentrancy-no-eth
+    function withdrawFromSFC(uint256 withdrawId)
+        external
+        onlyRegistrator
+        nonReentrant
+        returns (uint256 withdrawnAmount)
+    {
+        require(withdrawId < nextWithdrawId, "Invalid withdrawId");
+
+        // Can still withdraw even if the validator is no longer supported
+        // Load the withdrawal from storage into memory
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(!isWithdrawnFromSFC(withdrawId), "Already withdrawn");
+
+        uint256 sBalanceBefore = address(this).balance;
+
+        ISFC(sfc).withdraw(withdrawal.validatorId, withdrawId);
+
+        // Save state to storage
+        withdrawnAmount = address(this).balance - sBalanceBefore;
+        pendingWithdrawals -= withdrawal.undelegatedAmount;
+        withdrawals[withdrawId].undelegatedAmount = 0;
+
+        // Wrap Sonic (S) to Wrapped Sonic (wS)
+        IWrappedSonic(wrappedSonic).deposit{ value: withdrawnAmount }();
+
+        // Transfer the Wrapped Sonic (wS) to the Vault
+        // slither-disable-next-line unchecked-transfer unused-return
+        IERC20(wrappedSonic).transfer(vaultAddress, withdrawnAmount);
+
+        // withdrawal.undelegatedAmount & withdrawnAmount can differ in case of slashing
+        emit Withdrawn(
+            withdrawId,
+            withdrawal.validatorId,
+            withdrawal.undelegatedAmount,
+            withdrawnAmount
+        );
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /// @notice returns a bool whether a withdrawalId has already been withdrawn or not
+    function isWithdrawnFromSFC(uint256 withdrawId) public view returns (bool) {
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(withdrawal.validatorId > 0, "Invalid withdrawId");
+        return withdrawal.undelegatedAmount == 0;
+    }
+
+    /// @dev restake any pending validator rewards for all supported validators
+    function restakeRewards(uint256[] calldata validatorIds)
+        external
+        nonReentrant
+    {
+        uint256 totalRewards = 0;
+        for (uint256 i = 0; i < validatorIds.length; ++i) {
+            uint256 rewards = ISFC(sfc).pendingRewards(
+                address(this),
+                validatorIds[i]
+            );
+
+            if (rewards > 0) {
+                totalRewards += rewards;
+                ISFC(sfc).restakeRewards(validatorIds[i]);
+            }
+        }
+
+        // TODO use Deposit event or something else?
```

Comment:
Maybe we can add a bool flag `isRestake` to `Deposit` event indicating whether it is a normal deposit or restaking of the rewards

Reply:
I think we should drop the `Deposit` event from `restakeRewards`. I think we should leave `Deposit` for when assets are transferred from the Vault to the strategy.